apps/token/lib/token/ledger.ex

```diff
@@ -0,0 +1,100 @@
+defmodule Token.Ledger do
+  @moduledoc """
+  Abstraction around the blockchain. Only module that interacts with the
+  underlying blockchain
+  """
+
+  alias Token.{Ledger, Wallet, Transaction, Transaction.Verify}
+  alias Blockchain.Block
+
+  # transactions are sent to the blockchain as raw maps
+  defimpl Blockchain.Data, for: Map do
+    def hash(map), do: Ledger.map_to_tx(map).hash
+
+    def verify(map, chain) do
+      tx = Ledger.map_to_tx(map)
+      if tx.inputs == [["0", 0]] do
+        # coinbase transaction are automatically validated
+        :ok
+      else
+        Verify.verify_transaction(tx, Ledger.find_func(chain))
+      end
+    end
+  end
+
+  def write(%Transaction{} = tx), do: write(tx, Mix.env)
```

Comment:
Why you do this? I am not a big fun of doing code base on the environment. What you could do for this is to actually have plug in different modules on the `config` so you don't have to do this practices.

For example.
```elixir

# on config/config.exs
config :coinstack,
  ledger: CoinStack.Ledger

# on config/test.exs
config :coinstack,
  ledger: CoinStack.Test.Ledger # <----- this implements whatever should happen on the testing environment
```

Reply:
I totally agree with you. I do this because `Blockchain.add` is asynchronous (the transaction will be written in the blockchain only when a node mine the corresponding block), while `Blockchain.mine` is synchronous (the transaction is directly mined into a block, then the block is broadcasted to others).
But from the `Token` app point of view, the blockchain could be a simple list. So I will do a new module in the `Token` app: `Token.Test.Blockchain`, that will be a simple `agent` that stores an array of transactions and mimic the `Blockchain.add` API. The token system doesn't care about blocks, proof-of-work etc ... so that's the way to do it.

Reply:
Ok here is what I came up with: https://github.com/robinmonjo/blockchain/pull/7/commits/96732819f9a62890cadcbd5c0e9311736fa53822 
Let me know what you think

Reply:
much better ❤️  💟 💓